File: video_39_Buy_and_Sell_Stocks_With_Cooldown.java
import java.util.*;
class video_39_Buy_and_Sell_Stocks_With_Cooldown{
//method 1: memoization
    // public static long solve(int ind, int buy, long price[], long dp[][]){
    //     if(ind>=price.length)   return 0;
    //     if(dp[ind][buy] != -1) return dp[ind][buy];

    //     long profit = 0;
    //     if(buy == 1){
    //         profit = Math.max(-price[ind]+solve(ind+1, 0, price, dp),
    //         solve(ind+1, 1, price, dp));
    //     }
    //     else{
    //         profit = Math.max(price[ind]+solve(ind+2, 1, price, dp),
    //         solve(ind+1, 0, price, dp));
    //     }
    //     return dp[ind][buy] = profit;
    // }
    // public static long maximumProfit(long prices[], int n){
    //     //Code here
    //     long dp[][] = new long[n][2];
    //     for(int i = 0; i<n; i++)
    //         dp[i][0] = dp[i][1] = -1;
    //     return solve(0, 1, prices, dp);
    // }

//method 2: tabulation
    public static long maximumProfit(long price[], int n){
        //Code here
        long dp[][] = new long[n+2][2];
        dp[n][0] = dp[n][1] = dp[n+1][0] = dp[n+1][1] = 0;

        for(int ind = n-1; ind>=0; ind--){
            for(int buy = 0; buy<=1; buy++){
                long profit = 0;
                if(buy == 1){
                    profit = Math.max(-price[ind]+dp[ind+1] [0],
                    dp[ind+1][1]);
                }
                else{
                    profit = Math.max(price[ind]+dp[ind+2][1],
                    dp[ind+1][0]);
                }
                dp[ind][buy] = profit;
            }
        }
        return dp[0][1];
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        long arr[] = new long[n];
        for(int i = 0; i<n; i++)
            arr[i] = sc.nextInt();
        
        System.out.print(maximumProfit(arr, n));
    }
}

File: video_39_BuyAndSellStocksWithCooldown.cpp
// #include<iostream>
// #include<bits/stdc++.h>
// using namespace std;

/********************************************* MEMOIZATION **************************************************** */
// STOCKS WITH COOLDOWN EFFECT ie YOU CAN'T BUY STOCKS IMMEDIATELY AFTER SELLING PREVIOUS ONE

// int solve(int ind, vector<int> &arr, int buy, vector<vector<int>> &dp){
//     int n = arr.size();
//     if(ind >= n) return 0;
//     if(dp[ind][buy] != -1) return dp[ind][buy];
    // if(buy) return
    //     dp[ind][buy] = max(-arr[ind]+solve(ind+1, arr, 0, dp), solve(ind+1, arr, 1, dp));
    // else return
    //     dp[ind][buy] = max(arr[ind]+solve(ind+2, arr, 1, dp), solve(ind+1, arr, 0, dp));         
// }
// int stockProfit(vector<int> &prices){
//     // Write your code here.
//     int n = prices.size();
//     vector<vector<int>> dp(n, vector<int> (2, -1));
//     return solve(0, prices, 1, dp);
// }

/****************************************** TABULATION ******************************************************** */

// int stockProfit(vector<int> &arr){
//     // Write your code here.
//     int n = arr.size();
//     vector<vector<int>> dp(n+2, vector<int> (2, 0));//n+2 because we are here dealing with ind+2 also;
//     //No pint to write base case because the dp array is initialized to zero itself satisfying the base case
//     //dp[n-1][0] = 0; dp[n-1][1] = 0;
//     for(int ind = n-1; ind>=0; ind--){
//         for(int buy = 0; buy<=1; buy++){
//             if(buy == 1) 
//                 dp[ind][buy] = max(-arr[ind]+dp[ind+1][0], dp[ind+1][1]);
//             else
//                 dp[ind][buy] = max(arr[ind]+dp[ind+2][1], dp[ind+1][0]);
//         }
//     }
//     return dp[0][1]; 
// }

/****************************************** MAIN ************************************************************** */

// int main(){
//     int n;
//     cin>>n;
//     vector<int> arr(n, 0);
//     for(int i = 0; i<n; i++) cin>>arr[i];
//     cout<<stockProfit(arr);
//     return 0;
// }